XMonad/Actions/WindowNavigation.hs

```diff
@@ -135,87 +172,249 @@ swap = withTargetWindow swapWithFocused
+      io $ writeIORef stateRef $ modifyState oldWindowSet pos oldState
+
+-- | Get focused window and current position.
+getCurrentWindow :: Monad x => WNInput x -> x (Maybe (Window, Rectangle, Point))
+getCurrentWindow input@(_, oldWindowSet, _, _) =
+  whenJust' (pure $ W.peek oldWindowSet) Nothing $ \window -> do
+    (pos, rect) <- currentPosition input
+    return $ Just (window, rect, pos)
+
+-- | Gets the current position from the state passed in, or if nothing
+-- (say, from a restart), derives the current position from the current window.
+-- Also, verifies that the position is congruent with the current window
+-- (say, if you moved focus using mouse or something).
+-- Returns the window rectangle for convenience, since we'll need it later anyway.
+currentPosition :: Monad x => WNInput x -> x (Point, Rectangle)
+currentPosition (state, oldWindowSet, _, windowRect) = do
+  currentRect <- fromMaybe (Rectangle 0 0 0 0) <$> maybe (pure Nothing) windowRect (W.peek oldWindowSet)
+  let posMaybe = M.lookup (W.currentTag oldWindowSet) state
+      middleOf (Rectangle x y w h) = Point (midPoint x w) (midPoint y h)
+  return $ case posMaybe of
+    Nothing -> (middleOf currentRect, currentRect)
+    Just pos -> (centerPosition currentRect pos, currentRect)
+
+-- | Inserts new position into the state.
+modifyState :: WindowSet -> Point -> WNState -> WNState
+modifyState oldWindowSet =
+  M.insert (W.currentTag oldWindowSet)
+
+-- | "Jumps" the current position into the middle of target rectangle.
+-- (keeps the position as-is if it is already inside the target rectangle)
+centerPosition :: Rectangle -> Point -> Point
+centerPosition (Rectangle rx ry rw rh) pos@(Point x y) = do
+  if x >= rx && x < rx + fromIntegral rw && y >= ry && y < ry + fromIntegral rh
+  then pos
+  else Point (midPoint rx rw) (midPoint ry rh)
 
 midPoint :: Position -> Dimension -> Position
 midPoint pos dim = pos + fromIntegral dim `div` 2
 
-navigableTargets :: Point -> Direction2D -> X [(Window, Rectangle)]
-navigableTargets point dir = navigable dir point <$> windowRects
+-- | Make a list of target windows we can navigate to,
+-- sorted by desirability of navigation.
+navigableTargets :: Monad x => WNInput x -> Direction2D -> Rectangle -> Point -> x [(Window, Point)]
+navigableTargets input@(_, oldWindowSet, _, _) dir currentRect currentPos = do
+  allScreensWindowsAndRectangles <- mapSnd (rectTransform dir) <$> windowRects input
+  let
+    screenWindows = S.fromList $ W.integrate' $ W.stack $ W.workspace $ W.current oldWindowSet
+    (thisScreenWindowsAndRectangles, otherScreensWindowsAndRectangles) = partition (\(w, _) -> S.member w screenWindows) allScreensWindowsAndRectangles
+
+    pos = pointTransform dir currentPos
+    wr = rectTransform dir currentRect
+
+    rectInside r = (rect_p1 r >= rect_p1 wr && rect_p1 r < rect_p2 wr && rect_p2 r > rect_p1 wr && rect_p2 r <= rect_p2 wr) &&
```

Comment:
We also have https://hackage.haskell.org/package/xmonad-0.18.0/docs/XMonad-Operations.html#v:containedIn

Reply:
This one is less useful, because it doesn't handle the special case of a window that fully overlaps the current window in the orthogonal direction.